keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,389 @@
+
+### Graduation Criteria
+
+#### Alpha
+
+* Add feature flag for gating usage, off by default
+* Add flag to disable extra validation
+* Unit and integration tests
+
+#### Beta
+* Address user reviews and iterate if needed
+* Feature flag on by default
+
+#### GA
+* Remove feature flag
+
+### Upgrade / Downgrade Strategy
+
+Once feature flag is on by default (starting in Beta), administrators using custom serving certs
+can use the proposed flag to disable the extra validation and maintain current behavior.
+They will be able to use this flag even after the feature flag is removed.
+
+### Version Skew Strategy
+
+Not applicable.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate
+  - Feature gate name: `KubeletCertCNValidation`
+  - Components depending on the feature gate: kube-apiserver
+- [x] Other
+  - Describe the mechanism: kube-apiserver command-line flag `--disable-kubelet-cert-cn-validation`
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No. But requires restarting the kube-apiserver.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? No.
+
+###### Does enabling the feature change any default behavior?
+
+Yes. If a cluster is using custom kubelet serving certificates that don't follow the same convention as kubernetes issued certificates (CN is `system:node:<node-name>`),
+enabling this feature will make any connection initiated by the kube-api server fail (logs, exec and port-forwarding).
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, the feature can be disabled once enabled by just setting the command-line flag to true.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+You just get back the new behavior with the extra cert validation, no extra considerations needed.
+
+###### Are there any tests for feature enablement/disablement?
+
+We will add integration tests to validate the enablement/disablement flow. Test cases specified in a previous section.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+A rollout can fail if the feature flag is not enabled but the command-line flag is set.
+
+Already running workloads won't be impacted but cluster users won't be able to access the control plane if the cluster is single-node.
+
+###### What specific metrics should inform a rollback?
+
+Not applicable.
```

Comment:
The metric discussed above seems like it could be used since it would detect nodes with certs that do not have the correct CN.